### DISCUSSION TITLE
How to manage queryKey if it relies on the states from local component?

### DISCUSSION THREAD

**Discussion by s33h0w** [TanStack/query#3200] [Q&A] [answered]
React-query not only helps manage fetching data, but more importantly, it manages the result of the fetch request, i.e. the server state. It allows the server state to be available anywhere in the application, as long as the corresponding query key is known. I ran into a situation where the place that needs to get the query data doesn't always know the query key because it depends on the state of some local component. I would like to know what's the best practice in this case.

To better explain this situation, I have written a demo that contains two scenarios:
- the query relies on the state of a single local component
- the query relies on the states of multiple local components

In both cases, the query key relies on the local states. That is, these local states control a global server state. So, can these local states still be called "local"? Should they be elevated to a global state and managed by tools such as react-redux or react-context?

[online demo](https://codesandbox.io/s/react-query-query-with-local-state-b6ivr)

![image](https://user-images.githubusercontent.com/27037621/149651509-5aabacf0-9546-4940-9c55-e15721d08ae9.png)

**Comment by TkDodo** [comment]
> So, can these local states still be called "local"? Should they be elevated to a global state and managed by tools such as react-redux or react-context?

yes, this is a question of managing client state. If you need access to the client state in different components, either to display / edit it, or because its needed as an input for a query, you need to lift it up or make it more globally available. Often, keeping such data in the url is a good place because it will be globally available, you can subscribe to it and it will create sharable urls.

**Reply by s33h0w** [reply]
Thanks very much for your answer! I do agree with using some client management tools (e.g. react-redux) to manage these states or making them globally available as long as we need to access them in multiple different components. But about getting/setting query keys, I was wondering if it's possible not to use these methods above and just rely on react-query to achieve. Sometimes I can't put these data into the URL and also I don't want to introduce react-redux for them. What if we manage the query key inside the query function and return a method that can set the key value?

Below I assume an API `useQueryFactory` that takes `queryKeyReducer` as an argument instead of `queryKey`, and returns `dispatch` in addition to `queryResult`. In the component, I can initiate a new query by `dispatch`. This way, I don't have to raise some states to global, but still have the ability to change query key and generate a new query anywhere. Is the API `useQueryFactory` theoretically possible?

```typescript
// useGetXXX
import {useQueryFactory} from 'react-query';
// const {dispatch, ...queryResult} = useQueryFactory(queryKeyReducer, queryFn?, options?);

const useGetXXX = () => {
  const initialKey = 'XXX';
  const queryKeyReducer = (state, action) => {
  	switch (action.type) {
      case "param1":
        return {...state, param1: action.payload};
      case "param2":
        return {...state, param2: action.payload};
      default:
        return initialKey
    }
  }

  return userQueryFactory(queryKeyReducer, queryFn?, options?);
};

// Component
const Input = () => {
  const { dispatch, ...queryResult } = useGetXXX;
  const [param1, setParam1] = useState('');
  const handleChangeParam1 = (e) => {
    const value = e.target.value;
    setParam1(value);
    dispatch({
      type: 'param1',
      payload: value,
    });
  }
  
  return (
    <input value={param1} onChange={handleChangeParam1} />
  );
}
```

**Reply by zetsin** [reply]
This is a simple way with a global variable:

```
// useGetXXX
import {useQuery} from 'react-query';

const context = {
  request: null
}
const useGetXXX = (request) => {
  request = context.request = {
    ...context.request,
    ...request
  }
  return useQuery(["GetXXX", request], () => fetch(request), options?);
};

// Component
const Input = () => {
  const [param1, setParam1] = useState('');
  const queryResult = useGetXXX({
    param1
  });
  const handleChangeParam1 = (e) => {
    const value = e.target.value;
    setParam1(value);
  }
  
  return (
    <input value={param1} onChange={handleChangeParam1} />
  );
}
```

**Reply by s33h0w** [reply]
I think this may not work, as it will cause the changes that happened in one local component will not be synchronized with another component to get the latest data. This is because the query instances they subscribed won't have the same query key.

**Comment by s33h0w** [comment, marked as answer]
If some states of the local components can affect a globally shared server-side data managed by React Query, and we also don't want to lift those local states up to global. We can take a solution like this. Provide a globally shared queryKey, and return both the query result and a dispatch function that manages the global queryKey in the hook function. This way each local component has the ability to get and change the current query result. Here's what the code looks like. 

```typescript
export const [useSharedQueryKey, SharedQueryKeyProvider] = createReducerContext(xxx);
// query hook
export const useGetData = (): SharedQueryResult => {
    const [sharedQueryKey, dispatch] = useSharedQueryKey();

    return [
        useQuery(['XXXData', sharedQueryKey], () => fetchData(sharedQueryKey)),
        dispatch
    ];
};

// local component
export const LocalComponent = () => {
    const [value, setValue] = useState('');
    const [{data}, dispatch] = useGetData();

    const handleChange = (e) => {
        const value = e.target.value;
        setValue(value);
        dispatch(value);
    };

    return (
      <div>{data} <input onChange={handleChange} /> </div>
    );
}
```
[Live demo](https://codesandbox.io/s/react-query-shared-querykey-ounu0)

In the above solution I used [createReducerContext](https://github.com/streamich/react-use/blob/master/docs/createReducerContext.md) to manage the queryKey context. It is a factory for the react context and it's very simple you can write it yourself. Although we are still using react context to manage the global queryKey, it is harmless to the component's local states, and the component can still manage its local states itself. Going a step further, we can make the React Query provide the `useSharedQueryKey` so that we don't have to provide a context provider for each queryKey in the outer layer.
